### PR TITLE
Fix symlinks that are added to output package and points to not existing location

### PR DESCRIPTION
**Pull request by Fatme**
Fixes https://github.com/NativeScript/nativescript-cli/issues/584

This PR also fixes the pollution in package.json file, removes the need from `tmp` folder and provides a better way for doing incremental prepare of node_modules.

### CODE REVIEW

File: lib/tools/broccoli/broccoli.d.ts [hunk 1]
```diff
@@ -152,7 +152,7 @@ interface BroccoliNode {
 }
 
 interface IBroccoliBuilder {
-   prepareNodeModules(outputPath: string,  projectDir: string): IFuture<void>;
+   prepareNodeModules(outputPath: string,  projectDir: string, lastModifiedTime?: any): IFuture<void>;
```
Comment (teobugslayer) [inline lib/tools/broccoli/broccoli.d.ts]: The type should not be any but Date:
```
> fs.statSync("c:\\haxlogs.txt").mtime instanceof Date
true
```

File: lib/tools/broccoli/builder.ts [hunk 2]
```diff
@@ -5,29 +5,66 @@ let broccoli = require('broccoli');
...
 export class Builder implements IBroccoliBuilder {
+   private nodeModules: any = {};
+  
   constructor(private $fs: IFileSystem,
     private $nodeModulesTree: INodeModulesTree,
     private $projectDataService: IProjectDataService,
     private $logger: ILogger) { }
   
-  public prepareNodeModules(absoluteOutputPath: string, projectDir: string): IFuture<void> {
+  public prepareNodeModules(absoluteOutputPath: string, projectDir: string, lastModifiedTime?: any): IFuture<void> {
```
Comment (teobugslayer) [inline lib/tools/broccoli/builder.ts]: Again, type should be Date

File: lib/tools/broccoli/builder.ts [hunk 3]
```diff
@@ -5,29 +5,66 @@ let broccoli = require('broccoli');
...
-          this.rebuildNodeModulesTree(absoluteOutputPath, projectDir).wait();
-        }
-      } else { 
-         this.rebuildNodeModulesTree(absoluteOutputPath, projectDir).wait();
+      let nodeModulesPath = path.join(projectDir, "node_modules");
+            
+      if(lastModifiedTime) {      
+        let pipeline = gulp.src(path.join(projectDir, "node_modules/**"))
+          .pipe(vinylFilterSince(lastModifiedTime))
+          .pipe(through.obj( (chunk: any, enc: any, cb: any) => {
```
Comment (teobugslayer) [inline lib/tools/broccoli/builder.ts]: Type of cb is Function

File: lib/tools/broccoli/builder.ts [hunk 4]
```diff
@@ -5,29 +5,66 @@ let broccoli = require('broccoli');
...
+          }))
+          .pipe(gulp.dest(absoluteOutputPath));
+          
+         let future = new Future<any>();
+         
+         pipeline.on('end', (err: any, data: any) => {
+           if(err) {
+             future.throw(err);
+           } else {
+            future.return(data);
```
Comment (teobugslayer) [inline lib/tools/broccoli/builder.ts]: Why are you returning the data just to throw it away?

File: lib/services/platform-service.ts [hunk 5]
```diff
@@ -148,13 +148,20 @@ export class PlatformService implements IPlatformService {
...
+			this.$fs.ensureDirectoryExists(appDestinationDirectoryPath).wait();
 			var appSourceDirectoryPath = path.join(this.$projectData.projectDir, constants.APP_FOLDER_NAME);
-
+			
 			// Delete the destination app in order to prevent EEXIST errors when symlinks are used.
-			this.$fs.deleteDirectory(path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME)).wait();
-			shell.cp("-R", appSourceDirectoryPath, platformData.appDestinationDirectoryPath);
+			_(this.$fs.readDirectory(path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME)).wait())
+				.filter(directoryName => directoryName !== "tns_modules")
+				.each(directoryName => this.$fs.deleteDirectory(path.join(platformData.appDestinationDirectoryPath, "app", directoryName)).wait())
```
Comment (rosen-vladimirov) [inline lib/services/platform-service.ts]: "app" => constants.APP_FOLDER_NAME maybe?

File: lib/tools/broccoli/builder.ts [hunk 6]
```diff
@@ -5,29 +5,66 @@ let broccoli = require('broccoli');
...
+           }
+         });
+         
+         future.wait();
+      } else {
+         let nodeModulesDirectories = this.$fs.readDirectory(nodeModulesPath).wait();
+         _.each(nodeModulesDirectories, nodeModuleDirectoryName => {
+           let nodeModuleFullPath = path.join(nodeModulesPath, nodeModuleDirectoryName);
+           this.nodeModules[nodeModuleFullPath] = nodeModuleFullPath;
+         })
```
Comment (rosen-vladimirov) [inline lib/tools/broccoli/builder.ts]: missing `;`

File: lib/tools/broccoli/builder.ts [hunk 7]
```diff
@@ -5,29 +5,66 @@ let broccoli = require('broccoli');
...
       }
+      
+     let destCopy = $injector.resolve(destCopyLib.DestCopy, {
```
Comment (rosen-vladimirov) [inline lib/tools/broccoli/builder.ts]: isn't this the global injector? Maybe you should pass the injector in the constructor and use it from there.

File: lib/tools/broccoli/node-modules-dest-copy.ts [hunk 8]
```diff
@@ -14,23 +14,26 @@ import constants = require("./../../constants");
...
-	treeDiff.changedDirectories.forEach(changedDirectory => {
-		let changedDirectoryAbsolutePath = path.join(this.inputPath, constants.NODE_MODULES_FOLDER_NAME, changedDirectory);
+  constructor(private inputPath: string, private cachePath: string, private outputRoot: string, private projectDir: string) {
+	this.dependencies = Object.create(null);
+	this.devDependencies = this.getDevDependencies(projectDir);
+  }
+  
+  public rebuildChangedDirectories(changedDirectories: string[]): void {
+	_.each(changedDirectories, changedDirectoryAbsolutePath => {
+		//let changedDirectoryAbsolutePath = path.join(this.inputPath, constants.NODE_MODULES_FOLDER_NAME, changedDirectory);
```
Comment (rosen-vladimirov) [inline lib/tools/broccoli/node-modules-dest-copy.ts]: :smirk_cat:  commented code

File: lib/services/platform-service.ts [hunk 9]
```diff
@@ -148,13 +148,20 @@ export class PlatformService implements IPlatformService {
...
 
 			// Copy app folder to native project
+			this.$fs.ensureDirectoryExists(appDestinationDirectoryPath).wait();
 			var appSourceDirectoryPath = path.join(this.$projectData.projectDir, constants.APP_FOLDER_NAME);
-
+			
 			// Delete the destination app in order to prevent EEXIST errors when symlinks are used.
-			this.$fs.deleteDirectory(path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME)).wait();
-			shell.cp("-R", appSourceDirectoryPath, platformData.appDestinationDirectoryPath);
+			_(this.$fs.readDirectory(path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME)).wait())
```
Comment (rosen-vladimirov) [inline lib/services/platform-service.ts]: you can move `path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME)` outside of the chain and use the introduced variable in the path.join below

**Comment by dtopuzov** [comment]
:+1:

File: lib/tools/broccoli/builder.ts [hunk 10]
```diff
@@ -5,29 +5,81 @@ let broccoli = require('broccoli');
...
+         });
+         
+         future.wait();
+      }
+      
+      if(isNodeModulesModified && this.$fs.exists(absoluteOutputPath).wait()) {
+        let currentPreparedTnsModules = this.$fs.readDirectory(absoluteOutputPath).wait();
+        let tnsModulesInApp = this.$fs.readDirectory(path.join(projectDir, "app", "tns_modules")).wait();
+        let modulesToDelete = _.difference(currentPreparedTnsModules, tnsModulesInApp);
+        _.each(modulesToDelete, moduleName => this.$fs.deleteDirectory(path.join(absoluteOutputPath, moduleName)).wait())
```
Comment (rosen-vladimirov) [inline lib/tools/broccoli/builder.ts]: missing `;` at the end

**Comment by rosen-vladimirov** [comment]
:+1:  one missing `;`, but anyway - great work :100:

**Pull request merged**